c/cert/src/rules/DCL39-C/InformationLeakageAcrossTrustBoundariesC.ql

```diff
@@ -0,0 +1,22 @@
+/**
+ * @id c/cert/information-leakage-across-trust-boundaries-c
+ * @name DCL39-C: Avoid information leakage when passing a structure across a trust boundary
+ * @description Passing a structure with uninitialized fields or padding bytes can cause information
+ *              to be unintentionally leaked.
+ * @kind problem
+ * @precision medium
+ * @problem.severity error
+ * @tags external/cert/id/dcl39-c
+ *       security
+ *       external/cert/obligation/rule
+ */
+
+import cpp
+import codingstandards.c.cert
+import codingstandards.cpp.rules.informationleakageacrossboundaries.InformationLeakageAcrossBoundaries
+
+class InformationLeakageAcrossTrustBoundariesCQuery extends InformationLeakageAcrossBoundariesSharedQuery {
+  InformationLeakageAcrossTrustBoundariesCQuery() {
```

Comment:
I know this is an import BUT: as written I don't think this query is addressing the problem in the rule -- specifically, it doesn't appear to handle the cases it describes -- namely the edge cases in structure padding with it is non-homogeneous. 

Perhaps we should list out the limitations of this query in the implementation details?

```c

struct test {
  int a;
  char b;
  int c;
};
 
/* Safely copy bytes to user space */
extern int copy_to_user(void *dest, void *src, size_t size);
 
void do_stuff(void *usr_buf) {
  struct test arg = {.a = 1, .b = 2, .c = 3};
  copy_to_user(usr_buf, &arg, sizeof(arg));
}
```

Reply:
I agree an implementation note would be a good idea!

if I use the same testcase (the first NCE) in the [DCL55-CPP](https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL55-CPP.+Avoid+information+leakage+when+passing+a+class+object+across+a+trust+boundary) rule on the unmodified query that also does not get reported

since the related guidelines for DCL55-CPP points to DCL39-C and all of the examples are the same between them, and I believe the conditions are the same , I didnt spend too too much time digging into exactly what the CPP version was able to handle . but yeah now looking [closer I see ](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/trustboundary/UninitializedField.qll#L22) it only handles cases where the fields werent initialized by an initializer, moreso focuses on missing initialization than padding.

Reply:
(added the note!)

Reply:
Thanks @knewbury01 -- I think that's enough for now. As an additional step I wanted to suggest to open a issue to track this as an enhancement. I think that there is a large enough gap that we should probably keep an eye on it.

Reply:
added an enhancement issue as well!
https://github.com/github/codeql-coding-standards/issues/175